### PR TITLE
JDBC Flowable fixes

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.jboss.org/browse/ISPN-9915
https://issues.jboss.org/browse/ISPN-9916

### CODE REVIEW

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java [hunk 1]
```diff
@@ -543,10 +543,12 @@ public boolean contains(Object key) {
          }
          ps.setFetchSize(tableManager.getFetchSize());
          ResultSet rs = ps.executeQuery();
-         return function.apply(rs);
+         return function.apply(rs).doOnComplete(() -> JdbcUtil.safeClose(rs));
       }, kvp -> {
          JdbcUtil.safeClose(kvp.getValue());
-         connectionFactory.releaseConnection(kvp.getKey());
+         Connection connection = kvp.getKey();
+         connection.rollback();
```
Comment (wburns) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java]: We need to catch this exception, just to ensure we always return the connection to the pool. Probably just log it similarly to JdbcUtil#safeClose does.

**Comment by ryanemerson** [comment]
@wburns Fixed

**Comment by wburns** [comment]
Looks good, just want to let CI take a stab at it just to make sure.

**Comment by jesperpedersen** [comment]
`c.setAutoCommit(false)` and `c.rollback()` can't be called in transaction managed environments, such as WildFly.

Hence, the code should

```
   boolean ac = c.getAutoCommit();
   if (ac)
      c.setAutoCommit(false);

   ...

   if (ac)
      c.rollback();
```

HTH

**Comment by wburns** [comment]
@jesperpedersen thanks for the tip!

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java [hunk 2]
```diff
@@ -543,11 +539,39 @@ public boolean contains(Object key) {
...
+      final Connection connection;
+      final PreparedStatement statement;
+
+      FlowableConnection(ConnectionFactory factory, String sql) throws SQLException {
+         this.factory = factory;
+         this.connection = factory.getConnection();
+         this.autoCommit = connection.getAutoCommit();
+         this.statement = connection.prepareStatement(sql);
+
+         // Some JDBC drivers require auto commit disabled to do paging, but we must ensure autoCommit is true in case
```
Comment (wburns) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java]: I think this comment is slightly off. I think what it is is that the managed environment prevents these methods from being invoked by throwing an exception.

**Comment by wburns** [comment]
Changes seem good to me, just minor thing re: comments. Waiting for CI or @jesperpedersen if you have any other concerns.

**Comment by ryanemerson** [comment]
@wburns Good point, I have updated the comment

**Comment by jesperpedersen** [comment]
LGTM

**Comment by wburns** [comment]
Integrated into mater, thanks @ryanemerson !

**Pull request merged**